x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ForceMergeAction.java

```diff
@@ -92,6 +106,25 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         StepKey forceMergeKey = new StepKey(phase, NAME, ForceMergeStep.NAME);
         StepKey countKey = new StepKey(phase, NAME, SegmentCountStep.NAME);
 
+
+        if (this.bestCompression) {
+            StepKey closeKey = new StepKey(phase, NAME, CloseIndexStep.NAME);
+            StepKey openKey = new StepKey(phase, NAME, OpenIndexStep.NAME);
+            StepKey waitForGreenIndexKey = new StepKey(phase, NAME, WaitForIndexGreenStep.NAME);
+            StepKey updateCompressionKey = new StepKey(phase, NAME, UpdateSettingsStep.NAME);
+            Settings bestCompressionSettings = Settings.builder()
+                .put(EngineConfig.INDEX_CODEC_SETTING.getKey(), CodecService.BEST_COMPRESSION_CODEC).build();
+
+            CloseIndexStep closeIndexStep = new CloseIndexStep(closeKey, updateCompressionKey, client);
+            UpdateSettingsStep updateBestCompressionSettings = new UpdateSettingsStep(updateCompressionKey,
+                openKey, client, bestCompressionSettings);
+            OpenIndexStep openIndexStep = new OpenIndexStep(openKey, waitForGreenIndexKey, client);
+            WaitForIndexGreenStep waitForIndexGreenStep = new WaitForIndexGreenStep(waitForGreenIndexKey, forceMergeKey);
+            ForceMergeStep forceMergeStep = new ForceMergeStep(forceMergeKey, nextStepKey, client, maxNumSegments);
+            return Arrays.asList(closeIndexStep, updateBestCompressionSettings,
```

Comment:
I think this could iteratively build the list of steps and avoid having two paths, for example, in the best compression case, we miss the `readOnlyStep` since it has a different path.

Something like:

```java
List<Step> mergeSteps = new ArrayList<>();
mergeSteps.add(readOnlyStep);
if (this.bestCompression) {
    mergeSteps.add(closeIndexStep);
    mergeSteps.add(updateCompressionSettingsStep);
    mergeSteps.add(openIndexStep);
    mergeSteps.add(waitForGreenStep);
}
mergeSteps.add(forceMergeStep);
mergeSteps.add(segmentCountStep);

return mergeSteps;
```

(this is pseudocode and is missing a lot of the steps, but it's easier to follow I think)

Reply:
Totally agree on that. Just a question I might be wrong (not sure) since we would be unnecessary will be creating an object for the other flow and storing it. What you think?